chain/client/src/state_request_actor.rs

```diff
@@ -0,0 +1,237 @@
+use std::collections::VecDeque;
+use std::sync::Arc;
+
+use near_async::messaging::{Actor, Handler};
+use near_async::time::{Clock, Duration, Instant};
+use near_chain::state_sync::ChainStateSyncAdapter;
+use near_chain::types::RuntimeAdapter;
+use near_epoch_manager::EpochManagerAdapter;
+use near_network::client::{StateRequestHeader, StateRequestPart, StateResponse};
+use near_network::types::{StateResponseInfo, StateResponseInfoV2};
+use near_performance_metrics_macros::perf;
+use near_primitives::hash::CryptoHash;
+use near_primitives::state_sync::{
+    ShardStateSyncResponse, ShardStateSyncResponseHeader, ShardStateSyncResponseV3,
+};
+use near_store::adapter::chain_store::ChainStoreAdapter;
+use parking_lot::Mutex;
+use tracing::{error, info, warn};
+
+use crate::metrics;
+
+pub struct StateRequestActor {
+    clock: Clock,
+    state_sync_adapter: ChainStateSyncAdapter,
+    chain_store: ChainStoreAdapter,
+    genesis_hash: CryptoHash,
+    throttle_period: Duration,
+    num_state_requests_per_throttle_period: usize,
+    state_request_cache: Arc<Mutex<VecDeque<Instant>>>,
+}
+
+impl Actor for StateRequestActor {}
+
+impl StateRequestActor {
+    pub fn new(
+        clock: Clock,
+        runtime: Arc<dyn RuntimeAdapter>,
+        epoch_manager: Arc<dyn EpochManagerAdapter>,
+        genesis_hash: CryptoHash,
+        throttle_period: Duration,
+        num_state_requests_per_throttle_period: usize,
+    ) -> Self {
+        let chain_store = ChainStoreAdapter::new(runtime.store().clone());
+        let state_sync_adapter = ChainStateSyncAdapter::new(
+            clock.clone(),
+            chain_store.clone(),
+            epoch_manager,
+            runtime.clone(),
+        );
+        Self {
+            clock,
+            state_sync_adapter,
+            chain_store,
+            genesis_hash,
+            throttle_period,
+            num_state_requests_per_throttle_period,
+            state_request_cache: Arc::new(Mutex::new(VecDeque::default())),
```

Comment:
Nit: this can pre-allocate the exact amount of capacity for the `VecDeque` (`num_state_requests_per_throttle_period`) to avoid reallocating later on.